pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -302,6 +306,124 @@ func (m *kubeGenericRuntimeManager) getKubeletContainerStatuses(podSandboxID str
 	return statuses, nil
 }
 
+// generateContainerEvent generates an event for the container.
+func (m *kubeGenericRuntimeManager) generateContainerEvent(containerID kubecontainer.ContainerID, eventType, reason, message string) {
+	ref, ok := m.containerRefManager.GetRef(containerID)
+	if !ok {
+		glog.Warningf("No ref for container %q", containerID)
+		return
+	}
+	m.recorder.Event(ref, eventType, reason, message)
+}
+
+// executePreStopHook runs the pre-stop lifecycle hooks if applicable and returns new gracePeriod.
+func (m *kubeGenericRuntimeManager) executePreStopHook(pod *api.Pod, containerID kubecontainer.ContainerID, containerSpec *api.Container, gracePeriod int64) int64 {
+	start := unversioned.Now()
+	if pod != nil && containerSpec != nil && containerSpec.Lifecycle != nil && containerSpec.Lifecycle.PreStop != nil {
```

Comment:
Thanks. It's indeed much better to move the check to the caller, but I think keep `generateContainerEvent` inside this function is more clear. 

Made an update to addressing this (not a new commit so as to easy rebase other PRs), PTAL.